opencsp/app/sofast/SofastServer.py

```diff
@@ -0,0 +1,224 @@
+                            "slope_error": np.average(measurement.slopes_error),
+                            "slope_stddev": np.std(measurement.slopes_error),
+                        }
+                    )
+                else:
+                    ret["error"] = "Fringe measurement is not ready"
+                    ret["trace"] = "SofastServer.get_response::get_results_fringes"
+                    response_code = 409
+
+            else:
+                ret["error"] = f"Unknown action \"{action}\""
+                ret["trace"] = "SofastServer.get_response::N/A"
+                response_code = 404
+
+        except Exception as ex:
+            lt.error("Error in SofastServer with action " + action + ": " + repr(ex))
+            ret["error"] = (repr(ex),)
+            ret["trace"] = "".join(format_exception(ex))
+            response_code = 500
+
+        # sanity check: did we synchronize the error and response_code?
+        if response_code != 200:
+            if ret["error"] is None:
+                lt.error_and_raise(
+                    RuntimeError,
+                    f"Programmer error in SofastServer.get_response({action}): "
+                    + f"did not correctly set 'error' to match {response_code=}!",
+                )
+        if ret["error"] is not None:
+            if response_code == 200:
+                lt.error_and_raise(
+                    RuntimeError,
+                    f"Programmer error in SofastServer.get_response({action}): "
+                    + f"did not correctly set response_code to match {ret['error']=}!",
+                )
+
+        return response_code, json.dumps(ret)
+
+
+if __name__ == "__main__":
+    port = 8000
+
+    # Set up the logger
+    log_output_dir = opencsp_settings["sofast_server"]["log_output_dir"]
+    if log_output_dir is not None and ft.directory_exists(log_output_dir):
+        log_name_ext = "SofastServer_" + tdt.current_date_time_string_forfile() + ".log"
+        log_path_name_ext = os.path.join(log_output_dir, log_name_ext)
+        lt.logger(log_path_name_ext)
+
+    # Start the server
+    lt.warn(
+        "Warning in SofastServer: this server is unsecured. "
+        + f"It is suggested that you restrict outside access to port {port} of the host computer."
+    )
+    lt.info(f"Starting server on port {port}...")
+    server = HTTPServer(("0.0.0.0", port), SofastServer)
+
+    # Initialize the IO devices
+    ss.ServerState()
+    with ss.ServerState.instance() as state:
+        state.init_io()
+        state.load_default_settings()
+
+    # Lock in the currently allocated memory, to improve garbage collector performance
+    gc.collect()
+    gc.freeze()
+
+    # Start a new thread for the server
+    # The minimum time between server evaulation loops is determined by the GIL:
+    # https://docs.python.org/3/library/sys.html#sys.setswitchinterval
+    server_pool = ThreadPoolExecutor(max_workers=1)
+    server_pool.submit(server.serve_forever)
```

Comment:
What happens if this thread crashes or throws an exception? Will the program exit cleanly?

Reply:
This is a valid concern and actually why I'm doing this work. The intent is to have a separate program that restarts the entire process if the server stops responding.